front/lib/swr/data_source_view_documents.ts

```diff
@@ -0,0 +1,152 @@
+    isDocumentLoading: !disabled && !error && !data,
+    isDocumentError: error,
+    mutateDocument: mutate,
+  };
+}
+
+async function sendPatchRequest(
+  url: string,
+  {
+    arg,
+  }: {
+    arg: {
+      documentBody: PatchDataSourceWithNameDocumentRequestBody;
+    };
+  }
+) {
+  const res = await fetcherWithBody([url, arg.documentBody, "PATCH"]);
+  return res;
+}
+
+function decorateWithInvalidation<T>(
+  options: SWRMutationConfiguration<T, Error, string> | undefined,
+  invalidateCacheEntries: () => Promise<void>
+): SWRMutationConfiguration<T, Error, string> {
+  return options
+    ? {
+        ...options,
+        onSuccess: async (data, key, config) => {
+          await options.onSuccess?.(data, key, config);
+          await invalidateCacheEntries();
+        },
+      }
+    : {
+        onSuccess: invalidateCacheEntries,
+      };
+}
+
+export function useUpdateDataSourceViewDocument(
+  owner: LightWorkspaceType,
+  dataSourceView: DataSourceViewType,
+  documentName: string,
+  options?: SWRMutationConfiguration<PatchDocumentResponseBody, Error, string>
+) {
+  // Used only for cache invalidation
+  const { mutate: mutateContentNodes } = useDataSourceViewContentNodes({
+    owner,
+    dataSourceView,
+    disabled: true,
+  });
+
+  // Used only for cache invalidation
+  const { mutateDocument } = useDataSourceViewDocument({
+    owner,
+    dataSourceView,
+    documentId: documentName,
+    disabled: true,
+  });
+
+  // Decorate options's onSuccess with cache invalidation
+  const invalidateCacheEntries = async () => {
+    await Promise.all([mutateContentNodes, mutateDocument]);
+  };
+  const decoratedOptions = decorateWithInvalidation(
+    options,
+    invalidateCacheEntries
+  );
+
+  const patchUrl = `/api/w/${owner.sId}/spaces/${dataSourceView.spaceId}/data_sources/${dataSourceView.dataSource.sId}/documents/${documentName}`;
+  return useSWRMutation(patchUrl, sendPatchRequest, decoratedOptions);
+}
+
+async function sendPostRequest(
+  url: string,
+  {
+    arg,
+  }: {
+    arg: {
+      documentBody: PostDataSourceWithNameDocumentRequestBody;
+    };
+  }
+) {
+  const res = await fetcherWithBody([url, arg.documentBody, "POST"]);
+  return res;
+}
+
+export function useCreateDataSourceViewDocument(
+  owner: LightWorkspaceType,
+  dataSourceView: DataSourceViewType,
+  options?: SWRMutationConfiguration<PostDocumentResponseBody, Error, string>
+) {
+  // Used only for cache invalidation
+  const { mutate: mutateContentNodes } = useDataSourceViewContentNodes({
+    owner,
+    dataSourceView,
+    disabled: true,
+  });
+
+  // Decorate options's onSuccess with cache invalidation
+  const invalidateCacheEntries = async () => {
+    await mutateContentNodes();
+  };
+  const decoratedOptions = decorateWithInvalidation(
+    options,
+    invalidateCacheEntries
+  );
+
+  const createUrl = `/api/w/${owner.sId}/spaces/${dataSourceView.spaceId}/data_sources/${dataSourceView.dataSource.sId}/documents`;
```

Comment:
The term mutate is used ambiguously imho:
- 'mutate- in swr means 'invalidate' cache -> that's why I import mutate from above hooks
- but a mutation's not about changing the cache in the first place, it's about making a post/patch request

I pass this url because that's the one used by the mutation (2nd def)

Reply:
Yeah but SWR combine both IIRC, is there a case where we want to invalidate without re-triggering a request?

Reply:
That's correct, we also invalidate the post url (which has no effect because it's never used with GET anyway).
I don't think we'd ever want to invalidate without re-triggering a request in that case

Reply:
Seen IRL, it's using `useSWRMutation` here specifically for handling POST/PATCH requests, rather than using the regular `fetch` API that we use elsewhere in the codebase. This is the chosen method for mutation operations in this context. @overmode you own the part to refactor the other pattern since we don't want too many different patterns to achieve one same goal 🙃.

Reply:
@overmode Doubling down here that code consistency is extremely important in the long run. Did you create an issue to tracks that? Did you commit to an ETA?

2 things here:
 
- I think such changes of patterns probably require consensus from the engineering team as a whole and as such would be greatly helped by writing up a short design doc to get buy-in (vs pushing it in one place in one PR).
- It is important to always aim at having a codebase that "feels like" it was written by on person. Each small divergence from that is a potential big hit in productivity for engineers ramping up on the codebase or engineers visiting that small village that just does not look like the rest of the country.

As immediate next step I would suggest reverting to the simple pre-existing approaches. They're never perfect but they have the value of being used broadly. By doing so you save maybe 100s of people many minutes i the future looking at code they don't recognize.

From there you are more than welcome to write-up a small design doc to propose improvement to the existing and tackle a broad change so that we have one way of doing things.